CMakeLists.txt

```diff
@@ -48,75 +33,87 @@ LINK_DIRECTORIES(
+    trantor/net/inner/Connector.cc
+    trantor/net/inner/Poller.cc
+    trantor/net/inner/Socket.cc
+    trantor/net/inner/TcpConnectionImpl.cc
+    trantor/net/inner/Timer.cc
+    trantor/net/inner/TimerQueue.cc)
+
+FIND_PACKAGE(OpenSSL)
+IF(OpenSSL_FOUND)
+    add_definitions(-DUSE_OPENSSL)
     include_directories(${OPENSSL_INCLUDE_DIR})
     link_libraries(${OPENSSL_LIBRARIES})
-    AUX_SOURCE_DIRECTORY(trantor/net/ssl DIR_SRC)
-endif()
+    SET(trantor_sources ${trantor_sources} trantor/net/ssl/SSLConnection.cc)
+ENDIF()
 
 IF (CMAKE_SYSTEM_NAME MATCHES "Linux")
-    ADD_LIBRARY(trantor ${DIR_SRC} ${PROJECT_SOURCE_DIR}/trantor/net/inner/poller/EpollPoller.cc)
+    SET(trantor_sources ${trantor_sources} ${PROJECT_SOURCE_DIR}/trantor/net/inner/poller/EpollPoller.cc)
     MESSAGE(STATUS "current platform: Linux ")
 ELSEIF (CMAKE_SYSTEM_NAME MATCHES "Windows")
-    MESSAGE(STATUS "current platform: Windows")
+    MESSAGE(FATAL_ERROR "Error: Currently does not support Windows")
 ELSEIF (CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
-    ADD_LIBRARY(trantor ${DIR_SRC} ${PROJECT_SOURCE_DIR}/trantor/net/inner/poller/KQueue.cc)
+    SET(trantor_sources ${trantor_sources} ${PROJECT_SOURCE_DIR}/trantor/net/inner/poller/KQueue.cc)
     MESSAGE(STATUS "current platform: FreeBSD")
 ELSEIF(CMAKE_SYSTEM_NAME MATCHES "Darwin")
-    #ADD_LIBRARY(trantor ${DIR_SRC} ${PROJECT_SOURCE_DIR}/trantor/net/inner/poller/PollPoller.cc)
-    ADD_LIBRARY(trantor ${DIR_SRC} ${PROJECT_SOURCE_DIR}/trantor/net/inner/poller/KQueue.cc)
+    SET(trantor_sources ${trantor_sources} ${PROJECT_SOURCE_DIR}/trantor/net/inner/poller/KQueue.cc)
     MESSAGE(STATUS "current platform: MacOS")
 ELSE ()
     MESSAGE(STATUS "other platform: ${CMAKE_SYSTEM_NAME}")
 ENDIF (CMAKE_SYSTEM_NAME MATCHES "Linux")
 
+ADD_LIBRARY(trantor ${trantor_sources})
 
-SET(CONFIG_HEADER "${PROJECT_SOURCE_DIR}/config.h")
-file(WRITE "${CONFIG_HEADER}" "#pragma once\n")
-foreach(loop_var ${DEFS})
-    file(APPEND "${CONFIG_HEADER}" "#define ${loop_var}\n")
-endforeach(loop_var)
-file(APPEND "${CONFIG_HEADER}" "\n")
-
-if(DEFINED USE_STD_ANY)
-
-file(APPEND "${CONFIG_HEADER}" "#include <any>\n")
-file(APPEND "${CONFIG_HEADER}" "using std::any;\n")
-file(APPEND "${CONFIG_HEADER}" "using std::any_cast;\n\n")
-file(APPEND "${CONFIG_HEADER}" "#include <string_view>\n")
-file(APPEND "${CONFIG_HEADER}" "typedef std::string_view string_view;\n\n")
-
-else()
-
-file(APPEND "${CONFIG_HEADER}" "#include <experimental/any>\n")
-file(APPEND "${CONFIG_HEADER}" "using std::experimental::any;\n")
-file(APPEND "${CONFIG_HEADER}" "using std::experimental::any_cast;\n\n")
-file(APPEND "${CONFIG_HEADER}" "#include <experimental/string_view>\n")
-file(APPEND "${CONFIG_HEADER}" "typedef std::experimental::basic_string_view<char> string_view;\n\n")
-
-endif()
-
-EXEC_PROGRAM(${PROJECT_SOURCE_DIR}/update_config.sh ARGS "${CONFIG_HEADER} ${PROJECT_SOURCE_DIR}/trantor/utils/config.h")
-
-if (MAKETEST STREQUAL YES)
+IF (MAKETEST STREQUAL YES)
```

Comment:
This is a highly non-standard variable name for testing, is it something manually made?

If you want to create a proper `test` command for cmake then you should use the `ctest` functionality built into `cmake`.  It looks like all this does is build some programs that I'm guessing are manually run?  This is a good start for ctest, but ctest adds build commands for testing, timing reports, success runs, etc...  In essence if you setup a ctest setup in here then you will be able to run it just via `ctest --build-and-test . ./build_test --build-generator Ninja` (or whatever generator you want to use, I like ninja for its speed, it's so much faster than makefiles, though if you wanted makefiles then instead of `Ninja` you'd do `'Unix Makefiles'` instead, or whatever generator you want), or if it is just a normal build then you can just do `make test` and it will build and run the tests.

The basics to setup ctest in this file would be:
```cmake
enable_testing()
add_test(testname1 exename1 arg1 arg2 ...)
add_test(testname2 exename2 arg1 arg2 ...)
```
If you want to report results to a central cdash server you can replace `enable_testing()` with `include(CTest)` to set that up, but meh on that.

To actually build the programs that correspond to `exename1`/`exename2`/etc you just test if `BUILD_TESTING` is set, and if so then you build them like normal.  This variable is automatically defined if testing is requested.

Official `add_test` docs:  https://cmake.org/cmake/help/v3.15/command/add_test.html
`ctest` command docs:  https://cmake.org/cmake/help/latest/manual/ctest.1.html
I highly recommend this tutorial:  https://cgold.readthedocs.io/en/latest/tutorials/tests.html

Some other potentially relevant links from a quick google search:

* https://gitlab.kitware.com/cmake/community/wikis/doc/ctest/Testing-With-CTest
* http://bastian.rieck.me/blog/posts/2017/simple_unit_tests/
* https://a4z.bitbucket.io/blog/2018/05/17/Speed-up-your-test-cycles-with-CMake.html

Reply:
I put several unit tests(with **gtest** lib) into the test branch. And these tests in master branch are manually run as you guessed. 
I don't know much about test tools. Can CTest run the client and the server test at same time  automatically? 
Normalizing these tests seems like a big project. Can I do them in the next PR? ^.^